src/background/telemetry/exception-telemetry-listener.ts

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { BaseActionPayload } from 'background/actions/action-payloads';
+import { TelemetryEventHandler } from 'background/telemetry/telemetry-event-handler';
+import { Logger } from 'common/logging/logger';
+import * as TelemetryEvents from '../../common/extension-telemetry-events';
+import { UnhandledExceptionTelemetryData } from '../../common/extension-telemetry-events';
+
+export class ExceptionTelemetryListener {
+    constructor(private readonly telemetryEventHandler: TelemetryEventHandler) {}
+
+    public initialize(
+        logger: Logger,
+        extWindow: Window = window,
+        extConsole: Console = console,
+    ): void {
+        const sendExceptionTelemetry = (message: string, stackTrace?: string): void => {
+            const telemetry: UnhandledExceptionTelemetryData = { message, stackTrace };
+            const payload: BaseActionPayload = {
+                telemetry,
+            };
+
+            this.telemetryEventHandler.publishTelemetry(
+                TelemetryEvents.UNHANDLED_EXCEPTION,
+                payload,
+            );
+        };
+
+        // Catch top level synchronous errors
+        extWindow.onerror = function (
+            message: string,
+            source: string,
+            lineno: number,
+            colno: number,
+            error: Error = null,
+        ) {
+            sendExceptionTelemetry(message, error?.stack);
+            return false;
+        };
+
+        // Catch errors thrown in promises
+        extWindow.onunhandledrejection = function (event: PromiseRejectionEvent) {
+            sendExceptionTelemetry(event.reason);
+            return false;
+        };
+
+        // Catch errors written to console.error
+        const consoleError = extConsole.error;
+        extConsole.error = function (message?: any, ...optionalParams: any[]) {
+            const err = optionalParams[0] as Error;
+            if (message || err) {
+                sendExceptionTelemetry(message, err?.stack);
```

Comment:
I think it's a bit misleading to call the console.error and logger.error cases "unhandled exceptons" - they aren't necessarily unhandled exceptions. It would probably be good to be able to distinguish the source of the error (window error vs unhandled rejection vs logger error, etc) as part of the telemetry.